### PR TITLE
Task: WebRTC

### PR DESCRIPTION
**Pull request by lgrahl**
These changes add the WebRTC task protocol specification.

@dbrgn Please, review. :)

### CODE REVIEW

**Comment by lgrahl** [comment]
The discussed CSN changes haven't made it into the PR, I'll add those to the master directly.

**Comment by lgrahl** [comment]
Closes #45, closes #30, closes #28, closes #4, closes #3

File: Task-WebRTC.md [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+* A client SHALL validate that the *cookie_2* field's value contains
+  16 bytes and is different to the other client's cookie it currently 
+  uses, different to the client's current cookie and different to the 
+  client's upcoming cookie.
+* The *exclude* field MUST contain a list/an array of WebRTC data 
+  channel IDs (non-negative integers) that SHALL not be used for the 
+  signalling channel. The client MUST update its internal list of 
+  excluded data channel ids by new values of the other client's list. 
+  The resulting list MUST be stored sorted in ascending order.
```
Comment (dbrgn) [inline Task-WebRTC.md:90]: Isn't the storage and especially the storage layout an implementation detail?

File: Task-WebRTC.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+following items:
+
+* A client MUST set the *cookie_2* field to 16 cryptographically 
+  secure random bytes. It SHALL be different to the cookie the client 
+  currently uses, the other client currently uses and, for initiators, 
+  the *cookie_2* the responder has sent in its task's data.
+* The *exclude* field MUST contain a list/an array of WebRTC data 
+  channel ids (non-negative integers) that SHALL not be used for the 
+  signalling channel. This list MUST be available to be set from user 
+  applications that use specific data channel ids.
```
Comment (dbrgn) [inline Task-WebRTC.md:73]: Right now I can't remember why we need this blacklist. Can you elaborate?

File: Task-WebRTC.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+* The *max_size* field MUST be set to the value described by the 
+  *Detecting the Maximum Message Size* section.
```
Comment (dbrgn) [inline Task-WebRTC.md:75]: Maybe we should call this `max_packet_size`?

File: Task-WebRTC.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+signalling channel to a dedicated data channel:
+
+1. The client creates a new data channel on the `RTCPeerConnection` 
+   instance with the `RTCDataChannelInit` dictionary/object set 
+   containing only the following values:
+   * *ordered* SHALL be set to `true`,
+   * *protocol* SHALL be set to the same subprotocol that has been 
+     negotiated with the server,
+   * *negotiated* MUST be set to `true`, and
+   * *id* SHALL be set to by counting upwards from `0` and using the
```
Comment (dbrgn) [inline Task-WebRTC.md:151]: Typo: `to by`

File: Task-WebRTC.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+     first number that is NOT present in the internal list of excluded 
+     data channel ids (exchanged in the task's data).
+2. The newly created `RTCDataChannel` instance shall be wrapped by 
+   following the *Wrapped Data Channel* section.
+3. As soon as the data channel is *open*, the client SHALL linger for 
+   a minimum of one second. During this phase, the client SHALL accept 
+   messages from both the original signalling channel and the new 
+   signalling channel based on the wrapped data channel. Note that 
+   both channels require slightly different handling of the nonce/
+   header.
```
Comment (dbrgn) [inline Task-WebRTC.md:161]: The lingering is completely replaced by the 2-way `handover` messages, right?

File: Task-WebRTC.md [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+channel), the following changes MUST be applied:
+
+* All references to the *cookie* SHALL be interpreted as references to 
+  *cookie 2*.
+* Source and destination addresses SHALL NOT be set.
+* The data channel id MUST be set to the id of the data channel the 
+  message will be sent on.
+* A signalling channel that is being handed over SHALL continue using 
+  the overflow number and sequence number counters from the WebSocket-
+  based implementation.
```
Comment (dbrgn) [inline Task-WebRTC.md:228]: This is already specified in the section "Signalling Channel Handover". I'd leave it away in order not to confuse the reader.

File: Task-WebRTC.md [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+The same procedure as described in the
+[SaltyRTC protocol specification](./Protocol.md#receiving-a-signalling-message)
+SHALL be followed. However, for all messages that are being exchanged 
+over wrapped data channels (such as the handed over signalling 
+channel), the following changes MUST be applied:
+
+* All references to the *cookie* SHALL be interpreted as references to 
+  *cookie 2*.
+* Source and destination addresses SHALL NOT processed or validated.
```
Comment (dbrgn) [inline Task-WebRTC.md:240]: Source and destionation numbers are not contained, right? Then the instruction not to validate them is a bit strange?

File: Task-WebRTC.md [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+     |             offer             |
+     |------------------------------>|
+     |             answer            |
+     |<------------------------------|
+     |      candidate (n times)      |
+     |<----------------------------->|
+     |             close             |
+     |<----------------------------->|
+     |                               |
+```
```
Comment (dbrgn) [inline Task-WebRTC.md:284]: This has already been changed in the meantime, but the newer version should probably include more optional candidates after the andover.

File: Task-WebRTC.md [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+session key pair and the other client's session key pair.
+
+## 'candidate' Message
+
+Both clients MAY send ICE candidates at any time to each other. 
+Clients SHOULD bundle available candidates.
+
+A client who sends an ICE candidate SHALL set the *candidate* field to 
+a list/an array of dictionaries/objects where each dictionary/object 
+SHALL contain the following fields:
```
Comment (dbrgn) [inline Task-WebRTC.md:338]: The example that has been added in the meantime contains a map as value for `candidate`, not an array.

**Review by dbrgn — COMMENTED** [review]
Here's my (late) review :) Note that I actually looked at the current version, not at this outdated one.

File: Task-WebRTC.md (hunk 2 above)
Comment (lgrahl) [reply Task-WebRTC.md:73]: It's for applications that use static pre-negotiated data channel ids. We don't want to get in their way.

File: Task-WebRTC.md (hunk 3 above)
Comment (lgrahl) [reply Task-WebRTC.md:75]: I'm fine with both. If you want it changed, do it. :)

File: Task-WebRTC.md (hunk 2 above)
Comment (dbrgn) [reply Task-WebRTC.md:73]: Ah, ok.

File: Task-WebRTC.md (hunk 1 above)
Comment (lgrahl) [reply Task-WebRTC.md:90]: Agree. It's necessary to determine the lowest possible data channel id not excluded by both clients.

File: Task-WebRTC.md (hunk 3 above)
Comment (dbrgn) [reply Task-WebRTC.md:75]: Done :)

File: Task-WebRTC.md (hunk 1 above)
Comment (dbrgn) [reply Task-WebRTC.md:90]: Yes, but if the implementation wants to use a different approach, it should be able to do so.

**Comment by lgrahl** [comment]
Let's move over to https://github.com/saltyrtc/saltyrtc-meta/pull/70 for further reviews and discussion. I'll go through the remaining review here and add changes to that PR.

File: Task-WebRTC.md (hunk 3 above)
Comment (lgrahl) [reply Task-WebRTC.md:75]: Woohoo, merge conflict. :fireworks:

File: Task-WebRTC.md (hunk 1 above)
Comment (lgrahl) [reply Task-WebRTC.md:90]: Changed it in 9a44d2385e876b4f47ba06bf525c7d5a29f6d29d

File: Task-WebRTC.md (hunk 4 above)
Comment (lgrahl) [reply Task-WebRTC.md:151]: Resolved itself by 9a44d2385e876b4f47ba06bf525c7d5a29f6d29d

File: Task-WebRTC.md (hunk 5 above)
Comment (lgrahl) [reply Task-WebRTC.md:161]: Yes

File: Task-WebRTC.md (hunk 7 above)
Comment (lgrahl) [reply Task-WebRTC.md:240]: Agree. I'll mark them as _not to be used_.

Comment (lgrahl) [reply Task-WebRTC.md:240]: Done ea5d781

File: Task-WebRTC.md (hunk 6 above)
Comment (lgrahl) [reply Task-WebRTC.md:228]: Yeah, I agree, fixed in 6795921

File: Task-WebRTC.md (hunk 8 above)
Comment (lgrahl) [reply Task-WebRTC.md:284]: Fair enough, candidates added in 17b35f7

File: Task-WebRTC.md (hunk 9 above)
Comment (lgrahl) [reply Task-WebRTC.md:338]: Fixed in aaab1d2

File: Task-WebRTC.md (hunk 3 above)
Comment (dbrgn) [reply Task-WebRTC.md:75]: Rebase :)

**Pull request merged**